### PR TITLE
Python BCs

### PR DESCRIPTION
**Pull request by chleh**
Python code for BCs
================
* [Hertz contact](https://github.com/chleh/ogs6/blob/python-bcs-for-pr/Tests/Data/Mechanics/Linear/Python/hertz-contact-bc.py) // [docu](https://deploy-preview-2170--ogs.netlify.com/docs/benchmarks/python-bc/hertz-contact/hertz-contact/)
* [Laplace equation with BCs by given functions](https://github.com/chleh/ogs6/blob/python-bcs-for-pr/Tests/Data/Elliptic/square_1x1_GroundWaterFlow_Python/bcs_laplace_eq.py) // [docu](https://deploy-preview-2170--ogs.netlify.com/docs/benchmarks/python-bc/python-laplace-eq/)

Limitations (not solved in this PR)
===========================
* Only works if all physical fields have the same ansatz functions, i.e., not for HM.
* The normal traction BC is special and not covered by this PR.
* Constraint boundary conditions (#2145) cope with secondary variables (fluxes). Fluxes cannot be *passed to* the Python scripts by the code of this PR.

TODO
=====
* [x] finish Hertz contact test
* [x] test with nonlinear BC
* [x] test with multi-variable process
* [x] input file documentation
* [x] maybe fix possible Python version mismatch (presumably between VTK and OGS)
* [x] fix git lfs tricks (*.pnK)
* [x] squash

Review
======
* [ ] attribute used
* [x] VTK Python version mismatch
* [x] discuss NLSolver changes.
* [ ] Large tests & Python?

Issues to open
============
* [x] Maybe the two versions of deriveBCMap could be merged into one?
* [x] Python BC for normal traction
* [x] Python BC for mixed FEM ansatz functions
* [x] passing secondary variables to Python BC

~~***Don't review yet! Soon: review commit-wise.***~~

### CODE REVIEW

**Comment by chleh** [comment]
https://github.com/chleh/ogs6/blob/95e2e25c862e6979129e0143e36a7126886c2ed0/Tests/Data/Elliptic/square_1x1_GroundWaterFlow_Python/bcs_laplace_eq.py#L1-L48

**Comment by endJunction** [comment]
I'm afraid to ask, but 132kloc?

**Comment by chleh** [comment]
Yes. I copied OGS, and Python (2 and 3), and combined the two, and now there are Python BCs. :rofl: 
No. I assume its due to WIP ParaView state files. I'll remove them. Maybe I'll keep some of them to simplify the validation of (changed) reference solutions.

**Comment by bilke** [comment]
@chleh Jenkins unstable (yellow) means there are Doxygen warnings:

https://jenkins.opengeosys.org/job/ufz/job/ogs/job/PR-2170/21/warnings18Result/

**Comment by chleh** [comment]
@bilke, thanks for the hint. I almost started to be puzzled.

**Comment by chleh** [comment]
Strange now windows compilation tells me
> Unknown compiler version - please run the configure tests and report the results

https://jenkins.opengeosys.org/blue/organizations/jenkins/ufz%2Fogs/detail/PR-2170/27/pipeline/46

**Comment by bilke** [comment]
> Unknown compiler version - please run the configure tests and report the results

This comes from Boost, you have to ignore it...

File: ProcessLib/CMakeLists.txt [hunk 1]
```diff
@@ -37,4 +37,14 @@ endif()
 
 if(OGS_USE_PCH)
     cotire(ProcessLib)
+    if (OGS_USE_PYTHON)
+        # If Python is wrongly configured, in the Python headers the macro
+        # snprintf might be defined even though it's not needed. Precompiled headers
+        # cause that this macro definition is visible in all of ProcessLib.
+        # That might lead to clashes with jsoncpp. In particular the following
+        # compilation error was observed on MSVC:
```
Comment (endJunction) [inline ProcessLib/CMakeLists.txt]: I could imagine that msvc version and python version would be helpful...

**Comment by chleh** [comment]
Thanks for the fast response. Currently it seems like I cannot add any git lfs file. I tried with a fresh clone of my fork, added a png. Git push shows:
```
$ git push -u origin test-lfs 
Locking support detected on remote "origin". Consider enabling it with:
  $ git config lfs.https://gitlab.opengeosys.org/ogs/ogs.git/info/lfs.locksverify true
info: Uploading failed due to unsupported Content-Type header(s).                                              
info: Consider disabling Content-Type detection with:
info: 
info:   $ git config lfs.contenttype false
Uploading LFS objects:   0% (0/1), 130 KB | 0 B/s, done
...
```

Could anybody please check if you can add git LFS stuff to your own forks? Just check if it's my personal problem.

**Comment by chleh** [comment]
Thanks for the fast response. Currently it seems like I cannot add any git lfs file. I tried with a fresh clone of my fork, added a png. Git push shows:
```
$ git push -u origin test-lfs 
Locking support detected on remote "origin". Consider enabling it with:
  $ git config lfs.https://gitlab.opengeosys.org/ogs/ogs.git/info/lfs.locksverify true
info: Uploading failed due to unsupported Content-Type header(s).                                              
info: Consider disabling Content-Type detection with:
info: 
info:   $ git config lfs.contenttype false
Uploading LFS objects:   0% (0/1), 130 KB | 0 B/s, done
...
```

Could anybody please check if you can add git LFS stuff to your own forks? Just check if it's my personal problem.

File: ProcessLib/BoundaryCondition/PythonBoundaryCondition.cpp [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        auto val = _bc_data.bc_object->getDirichletBCValue(
+            t, {xs[0], xs[1], xs[2]}, boundary_node_id, primary_variables);
+        if (!_bc_data.bc_object->isOverriddenEssential())
+        {
+            DBUG(
+                "Method `getDirichletBCValue' not overridden in Python "
+                "script.");
+            return;
+        }
+        if (val.first)
```
Comment (endJunction) [inline ProcessLib/BoundaryCondition/PythonBoundaryCondition.cpp]: if (!val,first) return

File: ProcessLib/BoundaryCondition/PythonBoundaryCondition.cpp [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+    for (auto const* node : _bc_data.boundary_mesh.getNodes())
+    {
+        auto const boundary_node_id = node->getID();
+        auto const bulk_node_id = bulk_node_ids_map[boundary_node_id];
+
+        // gather primary variable values
+        primary_variables.clear();
+        auto const num_var = _dof_table_boundary->getNumberOfVariables();
+        for (int var = 0; var < num_var; ++var)
```
Comment (endJunction) [inline ProcessLib/BoundaryCondition/PythonBoundaryCondition.cpp]: check for HM

File: Applications/CLI/CMakeLists.txt [hunk 4]
```diff
@@ -5,6 +5,47 @@ target_link_libraries(ogs
...
+    # flags are provided to ensure high quality code generation. In contrast to the
+    # pybind11_add_module() command, the CMake interface library only provides the
+    # minimal set of parameters to ensure that the code using pybind11 compiles, but
+    # it does not pass these extra compiler flags (i.e. this is up to you).
+    # TODO: Enable further compiler/linker flags.
+    target_link_libraries(ogs PRIVATE pybind11::embed)
+
+    add_library(ogs_python_bindings STATIC ogs_python_bindings.cpp)
+    target_link_libraries(ogs_python_bindings PRIVATE pybind11::embed)
+    target_include_directories(ogs_python_bindings PRIVATE ${PYTHON_INCLUDE_DIRS})
```
Comment (endJunction) [inline Applications/CLI/CMakeLists.txt]: necessary?

File: Applications/CLI/CMakeLists.txt [hunk 5]
```diff
@@ -5,6 +5,47 @@ target_link_libraries(ogs
...
+    # Since pybind11 is a metatemplate library, it is crucial that certain compiler
+    # flags are provided to ensure high quality code generation. In contrast to the
+    # pybind11_add_module() command, the CMake interface library only provides the
+    # minimal set of parameters to ensure that the code using pybind11 compiles, but
+    # it does not pass these extra compiler flags (i.e. this is up to you).
+    # TODO: Enable further compiler/linker flags.
+    target_link_libraries(ogs PRIVATE pybind11::embed)
+
+    add_library(ogs_python_bindings STATIC ogs_python_bindings.cpp)
+    target_link_libraries(ogs_python_bindings PRIVATE pybind11::embed)
```
Comment (endJunction) [inline Applications/CLI/CMakeLists.txt]: Consider PUBLIC.

File: Applications/CLI/CMakeLists.txt (hunk 4 above)
Comment (chleh) [reply Applications/CLI/CMakeLists.txt]: Cleaned up.

File: Applications/CLI/CMakeLists.txt (hunk 5 above)
Comment (chleh) [reply Applications/CLI/CMakeLists.txt]: Cleaned up.

File: Applications/CLI/ogs_embedded_python.cpp [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+namespace ApplicationsLib
+{
+pybind11::scoped_interpreter setupEmbeddedPython()
+{
+#ifndef OGS_BUILD_SHARED_LIBS
+    // pybind11_init_impl_OpenGeoSys is the function initializing the embedded
+    // OpenGeoSys Python module. The name is generated by pybind11. If it is not
+    // obvious that this symbol is actually used, the linker might remove it
+    // under certain circumstances.
+    mark_used(&pybind11_init_impl_OpenGeoSys);
```
Comment (endJunction) [inline Applications/CLI/ogs_embedded_python.cpp]: just an idea...
Maybe declaring the `pybind11_init_impl_OpenGeoSys` function with `__attribute__((used))` helps (at least for gcc and clang).

File: NumLib/DOF/LocalToGlobalIndexMap.cpp [hunk 7]
```diff
@@ -249,7 +250,42 @@ LocalToGlobalIndexMap* LocalToGlobalIndexMap::deriveBoundaryConstrainedMap(
...
+    MeshLib::MeshSubset&& new_mesh_subset) const
+{
+    DBUG("Construct reduced local to global index map.");
+
+    // Create a subset of the current mesh component map.
+    std::vector<int> global_component_ids;
+
+    for (int i = 0; i < getNumberOfComponents(); ++i)
+    {
+        global_component_ids.push_back(i);
```
Comment (endJunction) [inline NumLib/DOF/LocalToGlobalIndexMap.cpp]: `std::iota(begin(global_component_ids), end(global_component_ids), getNumberOfComponents())`

File: NumLib/DOF/LocalToGlobalIndexMap.cpp [hunk 8]
```diff
@@ -249,7 +250,42 @@ LocalToGlobalIndexMap* LocalToGlobalIndexMap::deriveBoundaryConstrainedMap(
 
     return new LocalToGlobalIndexMap(
         std::move(all_mesh_subsets), global_component_ids,
-        _variable_component_offsets, elements, std::move(mesh_component_map));
+        _variable_component_offsets, elements, std::move(mesh_component_map),
+        ConstructorTag{});
+}
+
+std::unique_ptr<LocalToGlobalIndexMap>
+LocalToGlobalIndexMap::deriveBoundaryConstrainedMap(
```
Comment (endJunction) [inline NumLib/DOF/LocalToGlobalIndexMap.cpp]: Maybe the two versions of deriveBCMap could be merged into one?

File: ProcessLib/BoundaryCondition/Python/PythonBoundaryConditionLocalAssembler.h [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+        auto const num_nodes = _element.getNumberOfNodes();
+        auto const num_comp_total =
+            _data.dof_table_bulk.getNumberOfComponents();
+
+        auto const& bulk_node_ids_map =
+            *_data.boundary_mesh.getProperties().getPropertyVector<std::size_t>(
+                "bulk_node_ids");
+
+        // gather primary variables
+        Eigen::MatrixXd primary_variables_mat(num_nodes, num_comp_total);
```
Comment (endJunction) [inline ProcessLib/BoundaryCondition/Python/PythonBoundaryConditionLocalAssembler.h]: Optionally use compile time size for nodes' size...

File: Tests/Data/Mechanics/Linear/Python/gen-unit-circle.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    r2 = numpy_to_vtk(r2, 1)
+    r2.SetName("r2")
+
+    a = numpy_to_vtk(a, 1)
+    a.SetName("a")
+    # self.GetOutputDataObject(0).GetPointData().AddArray(a)
+    # self.GetOutputDataObject(0).GetPointData().AddArray(r2)
+    # self.GetOutputDataObject(0).GetPointData().AddArray(c2)
+
+    # self.GetOutputDataObject(0).SetPoints(pts) #.SetData(c2)
```
Comment (endJunction) [inline Tests/Data/Mechanics/Linear/Python/gen-unit-circle.py]: Maybe some debug comments could be removed...

File: NumLib/DOF/LocalToGlobalIndexMap.cpp (hunk 7 above)
Comment (chleh) [reply NumLib/DOF/LocalToGlobalIndexMap.cpp]: Indeed, verbose :smile:

File: ProcessLib/BoundaryCondition/Python/PythonBoundaryConditionLocalAssembler.h (hunk 9 above)
Comment (chleh) [reply ProcessLib/BoundaryCondition/Python/PythonBoundaryConditionLocalAssembler.h]: Hm... I think, I'd like to leave it all dynamic:
`Eigen::Matrix<double, ShapeFunction::size /*?*/, Eigen::Dynamic>` doesn't really look more pleasant.

File: ProcessLib/BoundaryCondition/Python/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+    PythonBoundaryConditionPythonSideInterface.h)
+
+target_compile_definitions(ProcessLibBoundaryConditionPython
+    PUBLIC OGS_USE_PYTHON)
+
+target_link_libraries(ProcessLibBoundaryConditionPython
+    PUBLIC BaseLib MathLib MeshLib NumLib logog)
+
+target_link_libraries(ProcessLibBoundaryConditionPython
+    PRIVATE pybind11::pybind11)
```
Comment (bilke) [inline ProcessLib/BoundaryCondition/Python/CMakeLists.txt]: You can merge this command with the previous one:

```cmake
target_link_libraries(ProcessLibBoundaryConditionPython
    PUBLIC BaseLib MathLib MeshLib NumLib logog
    PRIVATE pybind11::pybind11)
```

File: ProcessLib/BoundaryCondition/Python/PythonBoundaryCondition.cpp [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+        {
+            auto const num_comp =
+                _dof_table_boundary->getNumberOfVariableComponents(var);
+            for (int comp = 0; comp < num_comp; ++comp)
+            {
+                MeshLib::Location loc{_bc_data.bulk_mesh_id,
+                                      MeshLib::MeshItemType::Node,
+                                      bulk_node_id};
+                auto const dof_idx =
+                    _bc_data.dof_table_bulk.getGlobalIndex(loc, var, comp);
```
Comment (TomFischer) [inline ProcessLib/BoundaryCondition/Python/PythonBoundaryCondition.cpp]: Is it possible to compute the `dof_idx` also from the boundary dof table?

File: Tests/Data/Mechanics/Linear/Python/gen-unit-circle.py [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+
+grid = reader.GetOutput()
+assert grid.GetBounds() == (0.0, 1.0, 0.0, 1.0, 0.0, 0.0)
+
+coords = vtk_to_numpy(grid.GetPoints().GetData())
+
+a = np.empty(coords.shape[0])
+
+for i, (x, y, z) in enumerate(coords):
+    r = np.sqrt(x**2 + y**2)
```
Comment (TomFischer) [inline Tests/Data/Mechanics/Linear/Python/gen-unit-circle.py]: Where is `r` used?

File: ProcessLib/BoundaryCondition/Python/CMakeLists.txt (hunk 11 above)
Comment (chleh) [reply ProcessLib/BoundaryCondition/Python/CMakeLists.txt]: All pvsm files will be deleted.
Thanks for the hint. I'll change it.

File: ProcessLib/BoundaryCondition/Python/PythonBoundaryCondition.cpp (hunk 12 above)
Comment (chleh) [reply ProcessLib/BoundaryCondition/Python/PythonBoundaryCondition.cpp]: I think, it's not. I think I tried, but the dof table for the boundary is for a hypothetical equation system that is assembled solely on the boundary (i.e., with no connection to the bulk). Therefore all boundary dof indices would be wrong. @endJunction, is that right? If unsure, I can also check that.

File: Tests/Data/Mechanics/Linear/Python/gen-unit-circle.py (hunk 13 above)
Comment (chleh) [reply Tests/Data/Mechanics/Linear/Python/gen-unit-circle.py]: In some old version of this script, long, long time ago. I'll remove it.

File: ProcessLib/BoundaryCondition/Python/PythonBoundaryCondition.cpp [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+        LOGOG_COUT << std::flush;
+    }
+
+    //! Optionally flushes Python's stdout after running Python code.
+    ~FlushStdoutGuard()
+    {
+        if (!_flush)
+            return;
+
+        using namespace pybind11::literals;
```
[the remark this answers is not in the document]
Comment (chleh) [reply ProcessLib/BoundaryCondition/Python/PythonBoundaryCondition.cpp]: For the line below, e.g. `"end"_a` is the equivalent to a Python named argument.

File: ProcessLib/BoundaryCondition/Python/PythonBoundaryCondition.cpp [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+        pybind11::print("end"_a = "", "flush"_a = true);
```
[the remark this answers is not in the document]
Comment (chleh) [reply ProcessLib/BoundaryCondition/Python/PythonBoundaryCondition.cpp]: It's equivalent to (Python) `print(end="", flush=True)`, i.e., no end-of-line will be printed and output will be flushed.
So the behaviour is described in the docu of he destructor. Is that sufficient?

File: web/content/docs/benchmarks/python-bc/hertz-contact/hertz-contact.md [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+{{< data-link >}}
+
+## Problem description
+
+Two elastic spheres of same radius $R$ are brought into contact.
+The sphere centers are displaced towards each other by $w\_0$, with increasing
+values in every load step.
+Due to symmetry reasons a flat circular contact area of radius $a$ forms.
+
+{{< img src="../hertz-contact.pnK" >}}
```
[the remark this answers is not in the document]
Comment (chleh) [reply web/content/docs/benchmarks/python-bc/hertz-contact/hertz-contact.md]: The file names now are fixed.

File: NumLib/DOF/LocalToGlobalIndexMap.cpp (hunk 8 above)
Comment (chleh) [reply NumLib/DOF/LocalToGlobalIndexMap.cpp]: I'll open an issue.

File: ProcessLib/CMakeLists.txt (hunk 1 above)
Comment (chleh) [reply ProcessLib/CMakeLists.txt]: Issue solved. Code and comment removed.

File: Tests/Data/Mechanics/Linear/Python/gen-unit-circle.py (hunk 10 above)
Comment (chleh) [reply Tests/Data/Mechanics/Linear/Python/gen-unit-circle.py]: removed by now.

File: ProcessLib/BoundaryCondition/Python/PythonBoundaryConditionLocalAssembler.h [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+                _data.dof_table_bulk.getNumberOfVariableComponents(var);
+            for (int comp = 0; comp < num_comp; ++comp)
+            {
+                auto const global_component =
+                    dof_table_boundary.getGlobalComponent(var, comp);
+
+                for (unsigned element_node_id = 0; element_node_id < num_nodes;
+                     ++element_node_id)
+                {
+                    auto* node = _element.getNode(element_node_id);
```
[the remark this answers is not in the document]
Comment (chleh) [reply ProcessLib/BoundaryCondition/Python/PythonBoundaryConditionLocalAssembler.h]: changed.

File: ProcessLib/BoundaryCondition/Python/PythonBoundaryConditionLocalAssembler.h [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+                    auto const bulk_node_id =
+                        bulk_node_ids_map[boundary_node_id];
+                    MeshLib::Location loc{_data.bulk_mesh_id,
+                                          MeshLib::MeshItemType::Node,
+                                          bulk_node_id};
+                    auto const dof_idx =
+                        _data.dof_table_bulk.getGlobalIndex(loc, var, comp);
+                    if (dof_idx == NumLib::MeshComponentMap::nop)
+                    {
+                        // TODO extend Python BC to mixed FEM
```
[the remark this answers is not in the document]
Comment (chleh) [reply ProcessLib/BoundaryCondition/Python/PythonBoundaryConditionLocalAssembler.h]: `// TODO extend Python BC to mixed FEM ansatz functions` still fits on a single line (and therefore is displayed completely in QtCreator's todo list). I'd suggest tht wording.

File: Applications/CLI/ogs_embedded_python.cpp (hunk 6 above)
Comment (chleh) [reply Applications/CLI/ogs_embedded_python.cpp]: `__attribute__(used)` is gcc-specific. I didn't find the same for MSVC:
https://msdn.microsoft.com/de-de/library/74f4886t.aspx

Maybe `__declspec(dllexport)` might help?
https://msdn.microsoft.com/de-de/library/3y1sfaz2.aspx
But is that also applicable for static libs?

Anyway, we'd have to distinguish between different compilers, which is also kind of hackish...

**Comment by chleh** [comment]
@bilke, thanks for the VTK Python version hint.
`Modules/vtkPython.cmake` contains some info **(1)**:
```
set(VTK_PYTHONPATH "/usr/lib/python2.7/site-packages")
set(vtkPython_LIBRARIES "/usr/lib/libpython2.7.so")
```
Furthermore `/usr/lib/cmake/vtk-8.1/vtkModuleMacrosPython.cmake`
contains `find_package(PythonInterp ${VTK_PYTHON_VERSION} REQUIRED)`.
But probably that won't help us, because it just searches for some locally installed Python intepreter.

And this information is not propagated to OGS. My `CMakeCache.txt` contains basically only **(2)**
`VTK_DIR:PATH=/usr/lib64/cmake/vtk-8.1` as interesting VTK related information.

Taking (1) and (2) together, one could probably detect the Python version VTK is built with, if any. But is that really stable/portable?

File: ProcessLib/BoundaryCondition/Python/PythonBoundaryCondition.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (chleh) [reply ProcessLib/BoundaryCondition/Python/PythonBoundaryCondition.cpp]: Yes. `_a` is a [user-defined literal](https://en.cppreference.com/w/cpp/language/user_literal).
There are also other such literals, e.g. `s` in C++14.
E.g. in
```
using namespace std::string_literals;
auto str = "test"s;
```
`str` is a `std::string`, not a character array.

File: ThirdParty/CMakeLists.txt [hunk 19]
```diff
@@ -47,6 +47,33 @@ endif()
...
+        if (NOT "${vtkPython_LIBRARIES}" STREQUAL "${PYTHON_LIBRARIES}")
+            message(SEND_ERROR "Mismatch between VTK's and OpenGeoSys' Python "
+                "libraries: ${vtkPython_LIBRARIES} vs. ${PYTHON_LIBRARIES}. "
+                "This will lead to compilation or linking errors. "
+                "You can fix this error by using the same Python version for "
+                "OpenGeoSys as VTK is built with.")
+        endif()
+    endfunction()
+
+    check_python_version_compatibility()
```
Comment (bilke) [inline ThirdParty/CMakeLists.txt]: Looks good! If it works for you then I am fine. Jenkins will not complain as there are no VTK with Python binding packages used.

**Comment by chleh** [comment]
I just rebased and squashed this PR. Thank you for the quick reviewing process and for your help.
Lately, netlify failed because hugo could not resolve some page reference. Locally, at my machine it works, however. @bilke, could you please have a look.

Unfortunately I had to exclude the Hertz contact test, because it requires a non-trivial change to the non-linear solver, cf. #2184.

Remaining for this PR are:
* I decided not to use `__attribute__(used)` because it's not portable and I don't know of a version that works for MSVC.
* The Elder test case is a LARGE one. Are LARGE tests tested with a Python build?

If this PR builds successfully, from my perspective it's ready to be merged if you agree.

**Comment by endJunction** [comment]
LARGE tests are not tested.

**Comment by bilke** [comment]
For the failing web build merge this: https://github.com/chleh/ogs6/pull/10
